src/functions.php

```diff
@@ -42,48 +42,66 @@ function download(Backend $backend, $substitutes, callable $callback=null): arra
     return $backend->getVcards();
 }
 
+/**
+ * set up a stable FTP connection to a designated destination
+ * @param array $config
+ * @param string $pathKey array key to the destination path
+ * @param string $purpose part of error message
+ * @return stream
+ */
+function setUpFTPConnection($config, $pathKey, $purpose)
+{
+    $fritzbox = $config['fritzbox'];
+
+    $ftpserver = parse_url($fritzbox['url'], PHP_URL_HOST) ? parse_url($fritzbox['url'], PHP_URL_HOST) : $fritzbox['url'];
+    $connectFunc = (@$fritzbox['plainFTP']) ? 'ftp_connect' : 'ftp_ssl_connect';
+
+    if ($connectFunc == 'ftp_ssl_connect' && !function_exists('ftp_ssl_connect')) {
+        throw new \Exception("PHP lacks support for 'ftp_ssl_connect', please use `plainFTP` to switch to unencrypted FTP");
+    }
+
+    if (false === ($ftp_conn = $connectFunc($ftpserver))) {
+        $message = sprintf("Could not connect to ftp server %s for %s upload", $ftpserver, $purpose);
+        throw new \Exception($message);
+    }
+    if (!ftp_login($ftp_conn, $fritzbox['user'], $fritzbox['password'])) {
+        $message = sprintf("Could not log in %s to ftp server %s for %s upload", $fritzbox['user'], $ftpserver, $purpose);
+        throw new \Exception($message);
+    }
+    if (!ftp_pasv($ftp_conn, true)) {
+        $message = sprintf("Could not switch to passive mode on ftp server %s for s% upload", $ftpserver, $purpose);
+        throw new \Exception($message);
+    }
+    if (!ftp_chdir($ftp_conn, $fritzbox[$pathKey])) {
+        $message = sprintf("Could not change to dir %s on ftp server %s for %s upload", $fritzbox[$pathKey], $ftpserver, $purpose);
+        throw new \Exception($message);
+    }
+    return $ftp_conn;
+}
+
 /**
  * upload image files via ftp to the fritzbox fonpix directory
  *
  * @param stdClass[] $vcards downloaded vCards
  * @param array $config
- * @param array $phonebook
  * @param callable $callback
  * @return mixed false or [number of uploaded images, number of total found images]
  */
-function uploadImages(array $vcards, array $config, array $phonebook, callable $callback=null)
+function uploadImages(array $vcards, array $config, callable $callback=null)
 {
     $countUploadedImages = 0;
     $countAllImages = 0;
     $mapFTPUIDtoFTPImageName = [];                      // "9e40f1f9-33df-495d-90fe-3a1e23374762" => "9e40f1f9-33df-495d-90fe-3a1e23374762_190106123906.jpg"
     $timestampPostfix = substr(date("YmdHis"), 2);      // timestamp, e.g., 190106123906
 
-    if (null == ($imgPath = @$phonebook['imagepath'])) {
+    if (null == ($imgPath = @$config['phonebook']['imagepath'])) {
```

Comment:
gleiches Problem wie oben. Noch engere Koppelung als ohnehin schon. Wozu das?

Reply:
Hallo Andreas,
wie bereits geschrieben, möchte ich für eine meiner FeatureBranches die Codestrecke zum FTP-Aufbau wiederverwenden. 
Spricht grundsätzlich etwas dagegen, die CodeStrecke in eine eigene Function auszulagern?

Nach meinem Verständnis ist der Code verständlicher lesbar, wenn innerhalb der jeweiligen Funktionen auf die jeweiligen Bereiche der Konfiguration zugegriffen werden. Daher die Änderung in RunCommand bzw. im Funktionsaufruf:
- `$pictures = uploadImages($filtered, $this->config['fritzbox'], $this->config['phonebook'], function () use ($progress) {'
+ `$pictures = uploadImages($filtered, $this->config, function () use ($progress) {`
Wenn das gerade nicht gutem Coding entspricht, drehe ich es zurück.

Was ansonsten an Coding in `setUpFTPConnection` bzw. wenn dir der Name besser gefällt `getFtpConnection` steht, ist inhaltlich das gleiche was Wolfram programmiert hat - nur das ich es eben über Funktions-Parameter für die Wiederverwendung normalisiert habe (und die errorlog-Messages mit sprintf geändert habe (nach meinem Verständnis auch besser lesbar)
Darauf bezogen verstehe ich deine Anmerkung "Als bool $secure Parameter rausziehen." nicht - auch weil ich ehrlich gar nicht weiß, was Du damit meinst.

Grüße

Volker